service/worker/scheduler/spec.go

```diff
@@ -29,59 +29,129 @@ import (
 	"time"
 
 	"github.com/dgryski/go-farm"
-
 	schedpb "go.temporal.io/api/schedule/v1"
+
 	"go.temporal.io/server/common/primitives/timestamp"
 	"go.temporal.io/server/common/util"
 )
 
 type (
-	compiledSpec struct {
+	CompiledSpec struct {
 		spec     *schedpb.ScheduleSpec
 		tz       *time.Location
 		calendar []*compiledCalendar
 		excludes []*compiledCalendar
 	}
 )
 
-func newCompiledSpec(spec *schedpb.ScheduleSpec) (*compiledSpec, error) {
+func NewCompiledSpec(spec *schedpb.ScheduleSpec) (*CompiledSpec, error) {
+	// make shallow copy so we can change some fields
+	specCopy := *spec
+	spec = &specCopy
+
+	// parse CalendarSpecs to StructuredCalendarSpecs
+	for _, cal := range spec.Calendar {
+		structured, err := parseCalendarToStrucured(cal)
+		if err != nil {
+			return nil, err
+		}
+		spec.StructuredCalendar = append(spec.StructuredCalendar, structured)
+	}
+	spec.Calendar = nil
+
+	// parse ExcludeCalendars
+	for _, cal := range spec.ExcludeCalendar {
+		structured, err := parseCalendarToStrucured(cal)
+		if err != nil {
+			return nil, err
+		}
+		spec.ExcludeStructuredCalendar = append(spec.ExcludeStructuredCalendar, structured)
+	}
+	spec.ExcludeCalendar = nil
+
+	// parse CronStrings
+	for _, cs := range spec.CronString {
+		structured, interval, tz, err := parseCronString(cs)
+		if err != nil {
+			return nil, err
+		}
+		if tz != "" {
+			if spec.TimezoneName != tz || spec.TimezoneData != nil {
+				return nil, errConflictingTimezoneNames
+			}
+			spec.TimezoneName = tz
```

Comment:
no need this line, since we know TimezoneName must be equal to tz already

Reply:
oops, the conditional above is actually wrong. the intention is: a cron string is allowed to set tz if it's not set at the spec level. or if it is set at the spec level but is identical, then that's ok. otherwise error. but actually I need more cases if there are multiple cron strings.

clearly this needs more tests too.